examples/sudoku_zkp/src/main.rs

```diff
@@ -0,0 +1,254 @@
+use sunscreen::{
+    types::zkp::{ConstrainCmp, NativeField},
+    zkp_program, BackendField, Compiler, Runtime, ZkpProgramInput,
+};
+use sunscreen_zkp_backend::{bulletproofs::BulletproofsBackend, ZkpBackend};
+
+type BPField = NativeField<<BulletproofsBackend as ZkpBackend>::Field>;
+
+fn main() {
+    let app = Compiler::new()
+        .zkp_backend::<BulletproofsBackend>()
+        .zkp_program(sudoku_proof)
+        .compile()
+        .unwrap();
+
+    let prog = app.get_zkp_program(sudoku_proof).unwrap();
+
+    let runtime = Runtime::new_zkp(&BulletproofsBackend::new()).unwrap();
+
+    let ex_puzzle = [
+        [0, 7, 0, 0, 2, 0, 0, 4, 6],
+        [0, 6, 0, 0, 0, 0, 8, 9, 0],
+        [2, 0, 0, 8, 0, 0, 7, 1, 5],
+        [0, 8, 4, 0, 9, 7, 0, 0, 0],
+        [7, 1, 0, 0, 0, 0, 0, 5, 9],
+        [0, 0, 0, 1, 3, 0, 4, 8, 0],
+        [6, 9, 7, 0, 0, 2, 0, 0, 8],
+        [0, 5, 8, 0, 0, 0, 0, 6, 0],
+        [4, 3, 0, 0, 8, 0, 0, 7, 0],
+    ];
+
+    let ex_sol = [
+        [8, 7, 5, 9, 2, 1, 3, 4, 6],
+        [3, 6, 1, 7, 5, 4, 8, 9, 2],
+        [2, 4, 9, 8, 6, 3, 7, 1, 5],
+        [5, 8, 4, 6, 9, 7, 1, 2, 3],
+        [7, 1, 3, 2, 4, 8, 6, 5, 9],
+        [9, 2, 6, 1, 3, 5, 4, 8, 7],
+        [6, 9, 7, 4, 1, 2, 5, 3, 8],
+        [1, 5, 8, 3, 7, 9, 2, 6, 4],
+        [4, 3, 2, 5, 8, 6, 9, 7, 1],
+    ];
+
+    let board: Vec<ZkpProgramInput> = vec![ex_sol.map(|a| a.map(BPField::from)).into()];
+
+    let cons: Vec<ZkpProgramInput> = vec![ex_puzzle.map(|a| a.map(BPField::from)).into()];
+
+    let proof = runtime.prove(prog, cons.clone(), vec![], board).unwrap();
+
+    let verify = runtime.verify(prog, &proof, cons.clone(), vec![]);
+
+    assert!(verify.is_ok());
+}
+
+#[zkp_program(backend = "bulletproofs")]
+fn sudoku_proof<F: BackendField>(
+    #[constant] constraints: [[NativeField<F>; 9]; 9],
+    board: [[NativeField<F>; 9]; 9],
+) {
+    fn contains_circuit<F: BackendField>(
+        arr: [ProgramNode<NativeField<F>>; 9],
+        i: NativeField<F>,
+    ) -> ProgramNode<NativeField<F>> {
+        let mut circuit = NativeField::<F>::from(1).into_program_node();
+        for a in arr {
+            circuit = circuit * (i.into_program_node() - a);
+        }
+        return circuit;
+    }
+    // Proves that the board is actually valid: All inputs are from 1 to 9
+    let one = NativeField::<F>::from(1).into_program_node();
+    let nine = NativeField::<F>::from(9).into_program_node();
+    let zero = NativeField::<F>::from(0).into_program_node();
+
+    for i in 0..9 {
+        for j in 0..9 {
+            let square = board[i][j].into_program_node();
+            let constraint = constraints[i][j].into_program_node();
+            square.constrain_ge_bounded(one, 8);
+            square.constrain_le_bounded(nine, 8);
```

Comment:
It occurs to me that since we are ensuring the sets are `{1, ... , 9}` via that polynomial from `contains_circuit`, I don't think we need these two bounds checks ensuring `1 <= square <= 9`.

Reply:
You don't.